deepspeed/utils/comms_logging.py

```diff
@@ -139,6 +144,34 @@ def log_all(self):
                 avg_lat = trim_mean(vals[1], 0.1)
                 avg_algbw = trim_mean(vals[2], 0.1)
                 avg_busbw = trim_mean(vals[3], 0.1)
+                if print_log:
+                    print(
+                        f"{' ': <20}{convert_size(msg_size): <20}{count: <20}{total_lat: <20.2f}{avg_lat: <20.2f}{avg_algbw: <20.2f}{avg_busbw: <20.2f}"
+                    )
+
+        if show_straggler:
+            if print_log:
+                print("_______________________________")
+                print("Breakdown with straggler effect")
+                print("-------------------------------")
                 print(
-                    f"{' ': <20}{convert_size(msg_size): <20}{count: <20}{total_lat: <20.2f}{avg_lat: <20.2f}{avg_algbw: <20.2f}{avg_busbw: <20.2f}"
+                    f"{'Comm. Op': <20}{'Message Size': <20}{'Count': <20}{'Total comm lat(ms)': <20}{'Total straggler(ms)': <20}{'Avg comm lat(ms)': <20}{'Avg straggler(ms)': <20}"
                 )
+            for record_name in self.comms_dict.keys():
+                if print_log:
+                    print(record_name)
+                for msg_size, vals in sorted(self.comms_dict[record_name].items()):
+                    # vals[0] is the count for each msg size
+                    count = vals[0]
+                    # vals[1] is a list of latency records for each msg size
+                    lats = torch.tensor(vals[1])
+                    min_lats = torch.tensor(vals[1])
+                    dist.all_reduce(min_lats, op=ReduceOp.MIN)
```

Comment:
I am not convinced that this logic will compute straggler as defined. Since we are logging from rank 0 perspective, it seems this will report the difference between rank 0 and the fastest rank (i.e., with minimum latency). This will be fine if rank 0 latency is assumed to be greater or equal to other ranks. Is it okay to make this assumption?

Reply:
Straggler effect is the additional overhead introduced from syncing between different ranks.  Take a simple example, suppose a work load has two stages, each stage take 1s on average and has 0.3s average variance.

Assume there are two ranks:
Rank 0:
    stage 1: 0.8s
    stage 2: 1.4s
Total time = 2.2s
Rank 1:
    stage 1: 1.2s
    stage 2: 0.6s
Total time = 1.8s

If there is no syncing between two ranks, rank 0 will take 2.2s and rank 1 will take 1.8s.  If we run the workload for several times, we will observe rank 0 average runtime is 2.0s and rank 1 average runtime is 2.0s.  This is the base line we use to measure straggler effect.

If there is syncing between two ranks, assume allreduce take 0.1s, then at the end of each stage the fast rank will wait for slow rank.  So we will observe the following stage time of each rank:
Rank 0:
    stage 1: 1.3s --> allreduce time = 0.5s  = 0.4s (wait for another rank) + 0.1s (allreduce time itself)
    stage 2: 1.5s  --> allreduce time = 0.1s (allreduce time itself)
Total time = 2.8s, 0.6s longer than without syncing = 0.4s spending on waiting other rank + 0.2s allreduce time
Rank 1:
    stage 1: 1.3s --> allreduce time = 0.1s (allreduce time itself)
    stage 2: 1.5s --> allreduce time = 0.9s = 0.8s (wait for another rank) + 0.1s (allreduce time itself)
Total time = 2.8s, 1.0s longer than without synching = 0.8s spending on waiting other rank + 0.2s allreduce time

We can see that for each rank, the additional time spent when compared with no synching is the time waiting for other rank plus allreduce time itself.  The time waiting for other rank is the straggler effect of that rank.  If we run this setting for many times and observe a specific rank, average straggler effect will be 0.6s.

So its okay if rank 0 has smaller latency than other rank, in this specific allreduce, the straggler effect of rank 0 is zero, but in other allreduce, rank 0 may still has longer latency than some other ranks.  Straggler effect of rank 0 will reflect how much duration rank 0 spending on wait for other ranks, and this will statistically represent the straggler effect of other ranks.

Note straggler effect assume all accelerator has same performance but with variance/jitting introduced by random event such as cache miss, TLB miss, OS scheduling etc.   So statistically rank0 straggler effect can be used to study straggler effect of other ranks.

Reply:
Thanks for the detailed explanation. I think it is okay to measure straggler effect from perspective of rank 0. I was just thinking that it is possible to get largest straggler effect across the ranks by doing an extra allreduce. Something like 
```python
  local_straggeler_effect = sum(latency)-sum(allreduce(latency, OP=MIN))
  global_straggeler_effect = allreduce(local_straggler_effect, OP=MAX)
```
But I agree with your point that `local_stragger_effect` is sufficient over many iterations. I am fine with this now. Thanks!